ecs_pod_creator.go

```diff
@@ -236,6 +326,253 @@ func (e *EnvironmentVariable) SetSecretOptions(opts SecretOptions) *EnvironmentV
 	return e
 }
 
+// Validate checks that the environment variable name is given and that either
+// the raw environment variable value or the referenced secret is given.
+func (e *EnvironmentVariable) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(e.Name == nil, "must specify a name")
+	catcher.NewWhen(e.Value == nil && e.SecretOpts == nil, "must either specify a value or reference a secret")
+	catcher.NewWhen(e.Value != nil && e.SecretOpts != nil, "cannot both specify a value and reference a secret")
+	if e.SecretOpts != nil {
+		catcher.Wrap(e.SecretOpts.Validate(), "invalid secret options")
+	}
+	return catcher.Resolve()
+}
+
+// SecretOptions represents a secret with a name and value that may or may not
+// be owned by its pod.
+type SecretOptions struct {
+	PodSecret
+	// Exists determines whether or not the secret already exists or must be
+	// created before it can be used.
+	Exists *bool
+}
+
+// NewSecretOptions returns new uninitialized options for a secret.
+func NewSecretOptions() *SecretOptions {
+	return &SecretOptions{}
+}
+
+// SetName sets the secret name.
+func (s *SecretOptions) SetName(name string) *SecretOptions {
+	s.Name = &name
+	return s
+}
+
+// SetValue sets the secret value.
+func (s *SecretOptions) SetValue(val string) *SecretOptions {
+	s.Value = &val
+	return s
+}
+
+// SetExists sets whether or not the secret already exists or or must be
+// created.
+func (s *SecretOptions) SetExists(exists bool) *SecretOptions {
+	s.Exists = &exists
+	return s
+}
+
+// Validate validates that the secret name is given and that either the secret
+// already exists or the new secret's value is given
+func (s *SecretOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(s.Name == nil, "must specify a name")
+	catcher.NewWhen(!utility.FromBoolPtr(s.Exists) && s.Value == nil, "either a new secret's value must be given or the secret must already exist")
+	return catcher.Resolve()
+}
+
+// ECSPodExecutionOptions represent options to configure how a pod is started.
+type ECSPodExecutionOptions struct {
+	// Cluster is the name of the cluster where the pod will run. If none is
+	// specified, this will run in the default cluster.
+	Cluster *string
+	// PlacementOptions specify options that determine how a pod is assigned to
+	// a container instance.
+	PlacementOpts *ECSPodPlacementOptions
+	// SupportsDebugMode indicates that the ECS pod should support debugging, so
+	// you can run exec in the pod's containers. In order for this to work, the
+	// pod must have the correct permissions to perform this operation when it's
+	// defined. By default, this is false.
+	SupportsDebugMode *bool
+	// Tags are any tags to apply to the running pods.
+	Tags []string
```

Comment:
How are these tags different than the tags in `ESCPodCreateOptions`?

Reply:
The pod definition is basically a template (e.g. like a class) and the created pod is an instantiation of the template (e.g. like an object) - you have to register the template before you can make an instance of it. Since the template and the instance are considered separate resources within ECS, so they can have different tags.